### ISSUE TITLE
Performance issue when adding contract to the document

### ISSUE THREAD

**Issue by archhhh** [accordproject/template-studio-v2#160]
Description:

When I try to add any contract to the document in my first try it usually takes up to 6-7 seconds for any document to be added. in consequent attempts for the same document it usually takes less than 1 second.

To reproduce, just add a contract to the document from the list of templates in the TemplateLibrary component.

My machine specifications:
i5-7300HQ, 8 GB RAM

Browser:
- Google Chrome 71
- Mozilla Firefox 70

**Comment by archhhh** [comment]
I analyzed the performance once again with chrome developer tools and it revealed that the reason why it takes so long the first time it loads the contract is because it compiles the ergo contract to javascript client-side. 
![image](https://user-images.githubusercontent.com/26677794/69605386-4e519e80-1063-11ea-8415-eba76173423c.png)
For some contracts it may take up to 3-4 seconds. Don't you consider to move compiling part to the server to remove this delay on the client-end?

**Comment by jolanglinais** [comment]
Ah @archhhh this is great context. We have discussed which side these actions should take before, and I'm not completely clear on the logic behind client side.

@jeromesimeon @dselman @mttrbrts any thoughts?

**Comment by dselman** [comment]
Yes, thank you @archhhh for the analysis. At the moment TSv2 is *only* client side - so there's no obvious way to move this to server side. We may want to introduce a server at some point, but it will complicate deployment etc. Perhaps a web worker would be useful?

**Comment by mttrbrts** [comment]
Also, I'm surprised to see a latency this high for client-side compilation. Does this only happen on first load, or every compilation?

**Comment by jeromesimeon** [comment]
@archhhh Thanks for the detailed analysis. It's really useful. One place where we could compile-ahead would be when we build the template library, keeping both the source and a compiled version for JavaScript.

With that said, it's really quite long. I would need to dig a little further to know whether this is an "ecosystem" issue (the way the compiler itself is packaged for JavaScript) or a compiler issue (the actual complexity of the compiler).

Note that in TSv2, we would only need the full compilation to JavaScript to run tests, not when loading the template, nor when editing the model/logic/etc (which would only require the early compilation phases up to type-checking).

I'm not that familiar with Chrome's developers tools. Would someone be able to help me reproduce @archhhh 's analysis locally?

**Comment by jeromesimeon** [comment]
A few more thoughts:

@archhhh Would you be able to point me to templates for which the problem is particularly acute?

Also, in the user experience, I'm wondering if network time is a factor (how long it takes to obtain the template from it's URL).

Finally, are the redux numbers at the bottom are a little surprising too?

**Comment by jolanglinais** [comment]
> I'm not that familiar with Chrome's developers tools. Would someone be able to help me reproduce @archhhh 's analysis locally?

I'm unsure how @archhhh did it for that screenshot, but I think you can go to Performance in the dev tools, click record on the top left, add the clause template, and then stop recording once the template has fully loaded. It will show a lot of data.

EDIT: Then click on something and look at Event Log:
<img width="1668" alt="Screen Shot 2019-11-26 at 11 19 40 AM" src="https://user-images.githubusercontent.com/36460856/69651828-ad50ec80-103e-11ea-8d88-60c39fb21ee1.png">

**Comment by jeromesimeon** [comment]
Thanks @irmerk I think I found my way around a bit. The "Call Tree" does provide a nice aggregate (one you find the right "Microtask" at the top). It does look like actual compilation time to me, but I will have to investigate further to break it down within the `compileToJavaScript` call.

<img width="916" alt="Screenshot 2019-11-26 at 11 32 44 AM" src="https://user-images.githubusercontent.com/670099/69652994-80053e00-1040-11ea-8948-7fba2cd14723.png">

**Comment by archhhh** [comment]
> 
> 
> Also, I'm surprised to see a latency this high for client-side compilation. Does this only happen on first load, or every compilation?

@mttrbrts , Only on first load of a contract, subsequent loads of that same contract do not produce such delay

> 
> 
> I'm not that familiar with Chrome's developers tools. Would someone be able to help me reproduce @archhhh 's analysis locally?

@jeromesimeon https://developers.google.com/web/tools/chrome-devtools/evaluate-performance this can help you get started

> 
> 
> @archhhh Would you be able to point me to templates for which the problem is particularly acute?

It happens with each contract, the longer the size of the contract, the longer it takes to compile

> 
> 
> Also, in the user experience, I'm wondering if network time is a factor (how long it takes to obtain the template from it's URL).
> Finally, are the redux numbers at the bottom are a little surprising too?

I think redux-saga numbers are related to network time, and I think these numbers are alright.

**Comment by jeromesimeon** [comment]
> > Also, I'm surprised to see a latency this high for client-side compilation. Does this only happen on first load, or every compilation?
> 
> @mttrbrts , Only on first load of a contract, subsequent loads of that same contract do not produce such delay
> 
> > I'm not that familiar with Chrome's developers tools. Would someone be able to help me reproduce @archhhh 's analysis locally?
> 
> @jeromesimeon https://developers.google.com/web/tools/chrome-devtools/evaluate-performance this can help you get started
> 

Thanks!

> > @archhhh Would you be able to point me to templates for which the problem is particularly acute?
> 
> It happens with each contract, the longer the size of the contract, the longer it takes to compile
> 

That's actually a useful hint. We might be bumping again some earlier issue with internal string representation in the compiler.

> > Also, in the user experience, I'm wondering if network time is a factor (how long it takes to obtain the template from it's URL).
> > Finally, are the redux numbers at the bottom are a little surprising too?
> 
> I think redux-saga numbers are related to network time, and I think these numbers are alright.

Ok interesting. Thanks!
